File: dashboard/src/pages/Correlated.js
import React, { useState, useEffect } from 'react';
import { correlatedIncidents } from '../api';

export default function Correlated({ apiBase }) {
  const [list, setList] = useState([]);
  const [loading, setLoading] = useState(true);

  const load = (runNow = false) => {
    setLoading(true);
    correlatedIncidents(runNow, 24, 50)
      .then(setList)
      .catch((e) => { setList([]); console.error(e); })
      .finally(() => setLoading(false));
  };

  useEffect(() => {
    load();
  }, []);

  return (
    <div>
      <h1 className="page-title">Correlated Incidents</h1>
      <div className="card">
        <p className="muted">Alerts grouped by time proximity, topology, and metric similarity to reduce noise.</p>
        <button className="btn primary" onClick={() => load(true)}>Run correlation now (last 24h)</button>
        {loading && <p>Loading...</p>}
        {!loading && (
          <div>
            {list.map((g) => (
              <div key={g.group_id} className="card" style={{ marginTop: '0.5rem' }}>
                <h4>Group #{g.group_id}</h4>
                <p><strong>Summary:</strong> {g.root_cause_summary}</p>
                <p><strong>Affected nodes:</strong> {g.affected_nodes?.join(', ') || '—'}</p>
                <p><strong>Incident IDs:</strong> {g.incident_ids?.join(', ') || '—'}</p>
                {g.created_at && <p className="muted">Created: {new Date(g.created_at).toLocaleString()}</p>}
              </div>
            ))}
            {list.length === 0 && <p className="muted">No correlated groups. Click &quot;Run correlation now&quot; or wait for automatic correlation.</p>}
          </div>
        )}
      </div>
    </div>
  );
}

File: backend/routers/correlated.py
"""Event correlation API - GET /incidents/correlated."""
from typing import List
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from datetime import datetime, timedelta

from backend.schemas import CorrelatedGroupResponse
from backend.event_correlation.correlator import get_correlated_groups, run_correlation
from database.session import get_sync_session

router = APIRouter(prefix="/incidents", prefix_override="")


def get_db():
    db = get_sync_session()
    try:
        yield db
    finally:
        db.close()


@router.get("/correlated", response_model=List[CorrelatedGroupResponse])
def get_correlated_incidents(
    limit: int = Query(50, ge=1, le=200),
    run_now: bool = Query(False),
    hours: int = Query(24, ge=1, le=168),
    db: Session = Depends(get_db),
):
    """
    Return correlated incident groups (alerts grouped by time proximity, topology, metric similarity).
    Optionally run correlation now on last N hours.
    """
    if run_now:
        since = datetime.utcnow() - timedelta(hours=hours)
        run_correlation(db, since=since)
    raw = get_correlated_groups(db, limit=limit)
    return [
        CorrelatedGroupResponse(
            group_id=g["group_id"],
            root_cause_summary=g["root_cause_summary"],
            created_at=g["created_at"],
            incident_ids=g["incident_ids"],
            affected_nodes=g["affected_nodes"],
        )
        for g in raw
    ]
